crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1107,6 +1126,31 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    fn infer_type_alias_definition(
+        &mut self,
+        type_alias: &ast::StmtTypeAlias,
+        definition: Definition<'db>,
+    ) {
+        let type_alias_ty = Type::Todo;
+        let _span =
+            tracing::trace_span!("infer_type_alias_definition", type_alias=?type_alias).entered();
+
+        self.infer_expression(&type_alias.name);
+        if let Some(ref type_params) = type_alias.type_params {
+            self.infer_type_parameters(type_params);
+        }
+
+        self.infer_annotation_expression(&type_alias.value, DeferredExpressionState::Deferred);
+
+        // TODO: _with_binding?
```

Comment:
Yeah it should be `_with_binding` (as you currently have it), because the type statement defines a value for the type alias name at runtime.

The more questionable part is whether it should be a declaration, or just a binding. But I think it should be a declaration; we should complain if you reassign a type alias name to a different value.